### PR TITLE
[RISCV][CostModel] VPIntrinsics have same cost as their non-vp counterparts

### PR DESCRIPTION
**Pull request by michaelmaitland**
On RISCV, only a few VPIntrinsics have their cost modeled by the
VectorIntrinsicCostTable. Even so, none of those entries consider LMUL.
All other VPIntrinsics do not have meaningful modeling.

This patch models the cost of a VPIntrinsic as the cost of its non-VP
counterpart. It is possible that the VP Intrinsic is cheaper than the non-VP
version depending on VL. On RISCV, this may be due two reasons (if the
instruction is part of a loop):
  1. A smaller VL can be used on the last iteration of the loop.
  2. The VP instruction may avoid a scalar remainder loop.

I have left this as a TODO since I think this change puts us on the right path
of modeling the cost of a VPInstruction, and it isn't entierly clear to me how
much of a discount we should give to a known VL<VLMAX or what to do when VL
is unknown at compile time.

### CODE REVIEW

File: llvm/test/Analysis/CostModel/RISCV/gep.ll [hunk 1]
```diff
@@ -270,7 +270,7 @@ define void @non_foldable_vector_uses(ptr %base, <2 x ptr> %base.vec) {
 ; RVI-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %4 = getelementptr i8, ptr %base, i32 42
 ; RVI-NEXT:  Cost Model: Found an estimated cost of 5 for instruction: %x4 = call <2 x i8> @llvm.masked.expandload.v2i8(ptr %4, <2 x i1> undef, <2 x i8> undef)
 ; RVI-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %5 = getelementptr i8, ptr %base, i32 42
-; RVI-NEXT:  Cost Model: Found an estimated cost of 5 for instruction: %x5 = call <2 x i8> @llvm.vp.load.v2i8.p0(ptr %5, <2 x i1> undef, i32 undef)
+; RVI-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %x5 = call <2 x i8> @llvm.vp.load.v2i8.p0(ptr %5, <2 x i1> undef, i32 undef)
```
Comment (michaelmaitland) [inline llvm/test/Analysis/CostModel/RISCV/gep.ll:273]: This should probably call `getMemoryOpCost`

**Review by arcbbb — COMMENTED** [review]
LGTM. I'll await approval from others.
After this, it seems reasonable to pass VL to TTI cost function.

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h [hunk 2]
```diff
@@ -1687,6 +1687,60 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
...
+          VPIntrinsic::getFunctionalOpcodeForVP(ICA.getID());
+      if (FOp) {
+        // TODO: Support other kinds of Intrinsics (i.e. reductions)
+        if (ICA.getID() == Intrinsic::vp_load) {
+          Align Alignment;
+          if (auto *VPI = dyn_cast_or_null<VPIntrinsic>(ICA.getInst()))
+            Alignment = VPI->getPointerAlignment().valueOrOne();
+          unsigned AS = 0;
+          if (ICA.getArgs().size() > 1)
+            if (auto PtrTy = dyn_cast<PointerType>(ICA.getArgs()[0]->getType()))
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/BasicTTIImpl.h:1705]: `auto *PtrTy`

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h [hunk 3]
```diff
@@ -1687,6 +1687,60 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
...
+              AS = PtrTy->getAddressSpace();
+          return thisT()->getMemoryOpCost(*FOp, ICA.getReturnType(), Alignment,
+                                          AS, CostKind);
+        } else if (ICA.getID() == Intrinsic::vp_store) {
+          Align Alignment;
+          if (auto *VPI = dyn_cast_or_null<VPIntrinsic>(ICA.getInst()))
+            Alignment = VPI->getPointerAlignment().valueOrOne();
+          unsigned AS = 0;
+          if (ICA.getArgs().size() >= 2)
+            if (auto PtrTy = dyn_cast<PointerType>(ICA.getArgs()[1]->getType()))
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/BasicTTIImpl.h:1715]: `auto *PtrTy`

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h [hunk 4]
```diff
@@ -1687,6 +1687,60 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
...
+          Align Alignment;
+          if (auto *VPI = dyn_cast_or_null<VPIntrinsic>(ICA.getInst()))
+            Alignment = VPI->getPointerAlignment().valueOrOne();
+          unsigned AS = 0;
+          if (ICA.getArgs().size() > 1)
+            if (auto PtrTy = dyn_cast<PointerType>(ICA.getArgs()[0]->getType()))
+              AS = PtrTy->getAddressSpace();
+          return thisT()->getMemoryOpCost(*FOp, ICA.getReturnType(), Alignment,
+                                          AS, CostKind);
+        } else if (ICA.getID() == Intrinsic::vp_store) {
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/BasicTTIImpl.h:1709]: You can drop the else here

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h [hunk 5]
```diff
@@ -1687,6 +1687,60 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
...
+          Align Alignment;
+          if (auto *VPI = dyn_cast_or_null<VPIntrinsic>(ICA.getInst()))
+            Alignment = VPI->getPointerAlignment().valueOrOne();
+          unsigned AS = 0;
+          if (ICA.getArgs().size() >= 2)
+            if (auto PtrTy = dyn_cast<PointerType>(ICA.getArgs()[1]->getType()))
+              AS = PtrTy->getAddressSpace();
+          return thisT()->getMemoryOpCost(*FOp, Args[0]->getType(), Alignment,
+                                          AS, CostKind);
+        } else if (VPBinOpIntrinsic::isVPBinOp(ICA.getID())) {
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/BasicTTIImpl.h:1719]: You can drop the else here

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h [hunk 6]
```diff
@@ -1687,6 +1687,60 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
...
+      }
+
+      std::optional<Intrinsic::ID> FID =
+          VPIntrinsic::getFunctionalIntrinsicIDForVP(ICA.getID());
+      if (FID) {
+        // Non-vp version will have same Args/Tys except mask and vector length.
+        assert(ICA.getArgs().size() >= 2 && ICA.getArgTypes().size() >= 2 &&
+               "Expected VPIntrinsic to have Mask and Vector Length args and "
+               "types");
+        ArrayRef<const Value *> NewArgs(ICA.getArgs().begin(),
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/BasicTTIImpl.h:1732]: Can we do `ArrayRef(ICA.getArgs()).drop_back(2)`?

**Review by topperc — APPROVED** [review]
LGTM

**Comment by stellaraccident** [comment]
Hi there - we bisected to this change in a downstream compiler crash (on x86, I think due to the common codegen change). Here is the stack trace:

```
FAILED: tests/e2e/stablehlo_ops/check_llvm-cpu_local-task_reduce.mlir_module.vmfb /work/full-build-dir/tests/e2e/stablehlo_ops/check_llvm-cpu_local-task_reduce.mlir_module.vmfb 
cd /work/full-build-dir/tests/e2e/stablehlo_ops && /work/full-build-dir/tools/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=llvm-cpu --iree-input-type=stablehlo /work/tests/e2e/stablehlo_ops/reduce.mlir -o check_llvm-cpu_local-task_reduce.mlir_module.vmfb --iree-hal-executable-object-search-path=\"/work/full-build-dir\" --iree-llvmcpu-embedded-linker-path=\"/work/full-build-dir/llvm-project/bin/lld\" --iree-llvmcpu-wasm-linker-path=\"/work/full-build-dir/llvm-project/bin/lld\"
iree-compile: /work/third_party/llvm-project/llvm/include/llvm/Support/Casting.h:662: decltype(auto) llvm::dyn_cast(From *) [To = llvm::PointerType, From = llvm::Type]: Assertion `detail::isPresent(Val) && "dyn_cast on a non-existent value"' failed.
Please report issues to https://github.com/openxla/iree/issues and include the crash backtrace.
Stack dump:
0.	Program arguments: /work/full-build-dir/tools/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=llvm-cpu --iree-input-type=stablehlo /work/tests/e2e/stablehlo_ops/reduce.mlir -o check_llvm-cpu_local-task_reduce.mlir_module.vmfb --iree-hal-executable-object-search-path=\"/work/full-build-dir\" --iree-llvmcpu-embedded-linker-path=\"/work/full-build-dir/llvm-project/bin/lld\" --iree-llvmcpu-wasm-linker-path=\"/work/full-build-dir/llvm-project/bin/lld\"
 #0 0x00007fb3fd68375b llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /work/third_party/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:13
 #1 0x00007fb3fd6817b0 llvm::sys::RunSignalHandlers() /work/third_party/llvm-project/llvm/lib/Support/Signals.cpp:106:18
 #2 0x00007fb3fd683e3f SignalHandler(int) /work/third_party/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1
 #3 0x00007fb404b23420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #4 0x00007fb3f6c6f00b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #5 0x00007fb3f6c4e859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #6 0x00007fb3f6c4e729 (/lib/x86_64-linux-gnu/libc.so.6+0x22729)
 #7 0x00007fb3f6c5ffd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #8 0x00007fb40161f5dd (/work/full-build-dir/lib/libIREECompiler.so+0xa61f5dd)
 #9 0x00007fb40198d6f6 llvm::EVT::isSimple() const /work/third_party/llvm-project/llvm/include/llvm/CodeGen/ValueTypes.h:130:25
#10 0x00007fb40198d6f6 llvm::X86TTIImpl::getArithmeticReductionCost(unsigned int, llvm::VectorType*, std::optional<llvm::FastMathFlags>, llvm::TargetTransformInfo::TargetCostKind) /work/third_party/llvm-project/llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5078:10
#11 0x00007fb401995cee llvm::BasicTTIImplBase<llvm::X86TTIImpl>::getTypeBasedIntrinsicInstrCost(llvm::IntrinsicCostAttributes const&, llvm::TargetTransformInfo::TargetCostKind) /work/third_party/llvm-project/llvm/include/llvm/CodeGen/BasicTTIImpl.h:0:0
#12 0x00007fb4019888a0 llvm::BasicTTIImplBase<llvm::X86TTIImpl>::getIntrinsicInstrCost(llvm::IntrinsicCostAttributes const&, llvm::TargetTransformInfo::TargetCostKind) /work/third_party/llvm-project/llvm/include/llvm/CodeGen/BasicTTIImpl.h:0:0
#13 0x00007fb401988620 llvm::X86TTIImpl::getIntrinsicInstrCost(llvm::IntrinsicCostAttributes const&, llvm::TargetTransformInfo::TargetCostKind) /work/third_party/llvm-project/llvm/lib/Target/X86/X86TargetTransformInfo.cpp:4334:17
#14 0x00007fb40198980c llvm::BasicTTIImplBase<llvm::X86TTIImpl>::getIntrinsicInstrCost(llvm::IntrinsicCostAttributes const&, llvm::TargetTransformInfo::TargetCostKind) /work/third_party/llvm-project/llvm/include/llvm/CodeGen/BasicTTIImpl.h:1741:25
#15 0x00007fb401988620 llvm::X86TTIImpl::getIntrinsicInstrCost(llvm::IntrinsicCostAttributes const&, llvm::TargetTransformInfo::TargetCostKind) /work/third_party/llvm-project/llvm/lib/Target/X86/X86TargetTransformInfo.cpp:4334:17
#16 0x00007fb4018b0852 llvm::TargetTransformInfoImplCRTPBase<llvm::X86TTIImpl>::getInstructionCost(llvm::User const*, llvm::ArrayRef<llvm::Value const*>, llvm::TargetTransformInfo::TargetCostKind) /work/third_party/llvm-project/llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:1138:25
#17 0x00007fb403d5f3ed llvm::TargetTransformInfo::getInstructionCost(llvm::User const*, llvm::ArrayRef<llvm::Value const*>, llvm::TargetTransformInfo::TargetCostKind) const /work/third_party/llvm-project/llvm/lib/Analysis/TargetTransformInfo.cpp:266:3
#18 0x00007fb402bd7349 llvm::TargetTransformInfo::getInstructionCost(llvm::User const*, llvm::TargetTransformInfo::TargetCostKind) const /work/third_party/llvm-project/llvm/include/llvm/Analysis/TargetTransformInfo.h:411:12
#19 0x00007fb403b18388 llvm::InstructionCost::propagateState(llvm::InstructionCost const&) /work/third_party/llvm-project/llvm/include/llvm/Support/InstructionCost.h:57:19
#20 0x00007fb403b18388 llvm::InstructionCost::operator+=(llvm::InstructionCost const&) /work/third_party/llvm-project/llvm/include/llvm/Support/InstructionCost.h:100:5
#21 0x00007fb403b18388 llvm::CodeMetrics::analyzeBasicBlock(llvm::BasicBlock const*, llvm::TargetTransformInfo const&, llvm::SmallPtrSetImpl<llvm::Value const*> const&, bool) /work/third_party/llvm-project/llvm/lib/Analysis/CodeMetrics.cpp:180:14
#22 0x00007fb403265ebb llvm::FunctionSpecializer::run() /work/third_party/llvm-project/llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:0:0
#23 0x00007fb40325fbc3 runIPSCCP(llvm::Module&, llvm::DataLayout const&, llvm::AnalysisManager<llvm::Function>*, std::function<llvm::TargetLibraryInfo const& (llvm::Function&)>, std::function<llvm::TargetTransformInfo& (llvm::Function&)>, std::function<llvm::AssumptionCache& (llvm::Function&)>, std::function<llvm::DominatorTree& (llvm::Function&)>, std::function<llvm::BlockFrequencyInfo& (llvm::Function&)>, bool) /work/third_party/llvm-project/llvm/lib/Transforms/IPO/SCCP.cpp:165:5
#24 0x00007fb40325fbc3 llvm::IPSCCPPass::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /work/third_party/llvm-project/llvm/lib/Transforms/IPO/SCCP.cpp:402:8
#25 0x00007fb402ac373d llvm::detail::PassModel<llvm::Module, llvm::IPSCCPPass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /work/third_party/llvm-project/llvm/include/llvm/IR/PassManagerInternal.h:89:5
#26 0x00007fb4041f9330 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /work/third_party/llvm-project/llvm/include/llvm/IR/PassManager.h:521:10
#27 0x00007fb3fe7ca5e0 llvm::SmallPtrSetImplBase::isSmall() const /work/third_party/llvm-project/llvm/include/llvm/ADT/SmallPtrSet.h:195:33
```

We're going to revert locally. Especially given that this is impacting a target outside of RISCV, may I push a revert? We can try to generate a better repro asynchronously.

**Comment by michaelmaitland** [comment]
> may I push a revert? We can try to generate a better repro asynchronously.

Sure thing.

**Comment by stellaraccident** [comment]
> > may I push a revert? We can try to generate a better repro asynchronously.
> 
> Sure thing. Can you please show me how I can reproduce this upstream?

Verified that reverting this patch in our branch does fix the crash, so I'm considering this a strong indicator. I'm getting the recipe from my team for how to extract the LLVM-level reproducer.

**Comment by Groverkss** [comment]
> > may I push a revert? We can try to generate a better repro asynchronously.
> 
> Sure thing. Can you please show me how I can reproduce this upstream?

bug.ll : https://gist.github.com/Groverkss/4893d082222ea96dd6ac238cf8697cbd
`opt -O3 bug.ll`

This reproduces the bug for me

**Comment by michaelmaitland** [comment]
@stellaraccident @Groverkss thank you for pointing out the bug and providing me a way to reproduce upstream. I have posted https://github.com/llvm/llvm-project/pull/68752 which recommits this here with the bug fix included.

**Pull request merged**